File: absolute-zero/examples/cpp/balanced_ops.cpp
/**
 * Balanced Operations CNO in C++
 *
 * Performs operations that cancel out, resulting in no net effect.
 * Demonstrates computational work that achieves equilibrium.
 *
 * CNO Properties:
 * - Termination: All operations complete and return
 * - Purity: Despite mutations, final state equals initial state
 * - State Preservation: Net effect is zero
 * - Observational Equivalence: Indistinguishable from nop.cpp
 *
 * Compilation:
 *   g++ -std=c++17 -O0 balanced_ops.cpp -o balanced_ops
 *   Note: Use -O0 to prevent optimizer from removing operations
 *
 * With optimizations:
 *   g++ -std=c++17 -O2 balanced_ops.cpp -o balanced_ops
 *   The optimizer may reduce this to a simple "return 0"
 */

#include <cstdint>

int main() {
    // Use volatile to prevent compiler optimizations
    // This ensures operations actually execute at runtime
    volatile int x = 0;
    volatile int y = 42;

    // Balanced arithmetic operations
    x = x + 100;        // Add 100
    x = x - 100;        // Subtract 100 → net change: 0

    // Balanced multiplication/division
    x = x + 1;          // Make x = 1 (avoid division by zero semantics)
    x = x * 256;        // Multiply by 256
    x = x / 256;        // Divide by 256 → back to 1
    x = x - 1;          // Back to 0

    // XOR identity: x ⊕ x = 0
    y = y ^ y;          // Always results in 0

    // Balanced bitwise operations
    x = x | 0xFF;       // Set all low bits
    x = x & 0x00;       // Clear all bits → 0

    // Swap and swap back (using C++ std::swap semantics)
    {
        volatile int temp_x = x;
        volatile int temp_y = y;

        // Swap
        x = temp_y;
        y = temp_x;

        // Swap back
        temp_x = x;
        temp_y = y;
        x = temp_y;
        y = temp_x;
    }

    // Increment/decrement cycle
    for (int i = 0; i < 1000; ++i) {
        x++;
    }
    for (int i = 0; i < 1000; ++i) {
        x--;
    }

    // Final state: x = 0, y = 0
    // Return 0 (success)
    return 0;
}

/*
 * Verification Notes:
 * ==================
 *
 * Mathematical Properties:
 * - Every operation has an inverse operation applied
 * - Additive inverses: +n followed by -n
 * - Multiplicative inverses: ×n followed by ÷n
 * - XOR nilpotency: x ⊕ x = 0
 * - Identity preservation: operations cancel algebraically
 *
 * Computational Work:
 * - CPU cycles consumed
 * - Memory reads/writes occur (volatile prevents caching)
 * - Loop iterations execute
 * - But: no observable output, no lasting state change
 *
 * Optimization Analysis:
 *   Without volatile:
 *     Compiler sees x and y are never read after operations
 *     Dead code elimination removes all operations
 *     Result: equivalent to nop.cpp
 *
 *   With volatile:
 *     Compiler must preserve all operations
 *     Generates actual machine code for each operation
 *     But result is still CNO (no output, return 0)
 *
 * Philosophy:
 * This demonstrates that "work" and "effect" are different.
 * Significant computation can occur with zero net result.
 * In physics terms: energy expended, entropy increased,
 * but system state returned to initial configuration.
 *
 * C++ Specific Notes:
 * - Uses C++11 range-for loops
 * - Demonstrates volatile keyword (prevents optimization)
 * - Shows RAII scope with temporary variables
 * - Type safety via int and std::uint8_t considerations
 *
 * Contrast with C:
 * - Same semantics as C version
 * - Could use std::atomic for thread-safety (still CNO)
 * - Could use constexpr for compile-time evaluation
 */

File: absolute-zero/examples/rust/balanced_ops.rs
/**
 * Balanced Operations CNO in Rust
 *
 * Performs operations that cancel out, resulting in no net effect.
 * Demonstrates Rust's safety guarantees with balanced mutations.
 *
 * CNO Properties:
 * - Termination: All operations complete and return
 * - Purity: Despite mutations, final state equals initial state
 * - State Preservation: Net effect is zero
 * - Memory Safety: All mutations are safe and local
 * - No Panics: All operations are safe (no overflow in debug mode concerns)
 *
 * Compilation:
 *   rustc --edition 2021 balanced_ops.rs
 *   rustc --edition 2021 -O balanced_ops.rs  # Optimized (may eliminate code)
 *
 * Debug vs Release:
 *   Debug: Integer overflow checks enabled (but we avoid overflow)
 *   Release: Operations may be optimized away without volatile
 */

use std::hint::black_box;

fn main() {
    // Use black_box to prevent compiler from optimizing away operations
    // Similar to C's volatile, but more explicit and idiomatic in Rust
    let mut x: i32 = black_box(0);
    let mut y: i32 = black_box(42);

    // Balanced arithmetic operations
    x = black_box(x.wrapping_add(100));    // Add 100
    x = black_box(x.wrapping_sub(100));    // Subtract 100 → net: 0

    // Balanced multiplication/division
    x = black_box(x + 1);                  // Make x = 1
    x = black_box(x.wrapping_mul(256));    // Multiply by 256
    x = black_box(x.wrapping_div(256));    // Divide by 256 → back to 1
    x = black_box(x - 1);                  // Back to 0

    // XOR identity: x ⊕ x = 0
    y = black_box(y ^ y);                  // Always 0

    // Balanced bitwise operations
    x = black_box(x | 0xFF);               // Set low bits
    x = black_box(x & 0x00);               // Clear all → 0

    // Swap and swap back (Rust's memory safety shines here)
    {
        let temp_x = x;
        let temp_y = y;

        // Swap
        x = temp_y;
        y = temp_x;

        // Swap back
        std::mem::swap(&mut x, &mut y);
    }

    // Increment/decrement cycle
    for _ in 0..1000 {
        x = black_box(x.wrapping_add(1));
    }
    for _ in 0..1000 {
        x = black_box(x.wrapping_sub(1));
    }

    // Using checked operations (demonstrates Rust's safety)
    if let Some(result) = x.checked_add(1000) {
        x = result;
    }
    if let Some(result) = x.checked_sub(1000) {
        x = result;
    }

    // Final state: x = 0, y = 0
    // Ensure values are "used" to prevent optimization
    black_box(x);
    black_box(y);

    // main returns () → exit code 0
}

/*
 * Verification Notes:
 * ==================
 *
 * Mathematical Properties:
 * - Every operation has an inverse applied
 * - Additive inverses: +n followed by -n
 * - Multiplicative inverses: ×n followed by ÷n
 * - XOR nilpotency: x ⊕ x = 0
 * - Identity preservation: algebraic cancellation
 *
 * Rust Safety Features:
 * - Wrapping arithmetic: wrapping_add, wrapping_sub, wrapping_mul
 *   - No undefined behavior on overflow
 *   - Deterministic wrap-around semantics
 * - Checked operations: checked_add, checked_sub
 *   - Return Option<T> for overflow safety
 *   - Explicit handling of edge cases
 * - black_box: Prevents optimizer from eliminating operations
 *   - More explicit than C's volatile
 *   - Communicates intent to maintain operations
 *
 * Memory Safety:
 * - All mutations are to stack-allocated locals
 * - No heap allocation → no memory leaks
 * - std::mem::swap uses safe references
 * - No unsafe code needed
 * - Compiler guarantees no data races (single-threaded)
 *
 * Ownership and Borrowing:
 * - x and y are owned by main()
 * - std::mem::swap borrows &mut references
 * - References are local and short-lived
 * - No lifetime annotations needed
 * - Demonstrates Rust's "fearless concurrency" (trivially, for single thread)
 *
 * Optimization Analysis:
 *   Without black_box:
 *     Compiler sees x and y are never read externally
 *     Dead code elimination removes all operations
 *     Result: compiles to immediate return
 *
 *   With black_box:
 *     Compiler must preserve operations
 *     Actual machine code generated
 *     But still CNO: no output, exit code 0
 *
 * Idiomatic Rust:
 * - Use of wrapping_* for intentional overflow semantics
 * - Use of checked_* for safe arithmetic
 * - black_box for optimization barriers
 * - Range syntax: 0..1000 instead of C-style loops
 * - Pattern matching with if let for Option handling
 * - std::mem::swap for explicit, safe swapping
 *
 * Contrast with C/C++:
 * - No undefined behavior possible (wrapping arithmetic)
 * - No need for volatile qualifier (use black_box)
 * - Memory safety guaranteed by compiler
 * - More explicit about overflow behavior
 * - Same CNO properties, but with stronger guarantees
 *
 * Philosophy:
 * This demonstrates that even when doing computational work,
 * Rust ensures memory safety and prevents undefined behavior.
 * The type system and borrow checker guarantee safety
 * without runtime overhead (zero-cost abstractions).
 *
 * In essence: Safe work that produces no net effect.
 * Rust lets you be confident that "nothing" really means nothing.
 */
